ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -0,0 +1,441 @@
+ -  | muxDuplex | | muxDuplex | | muxDuplex | | muxDuplex | dedicated thread which will
+ -  | Initiator | | Responder | | Initiator | | Responder | send ByteStrings of CBOR encoded
+ -  | ChainSync | | ChainSync | | BlockFetch| | BlockFetch| data.
+ -  +-----+-----+ +-----+-----+ +-----+-----+ +-----+-----+
+ -        |             |             |             |
+ -         \            \            /             /
+ -          -------------------+-------------------
+ -                             | CBOR data
+ -                             V
+ -                           |  | For a given MuxBearer there is a single egress queue shared
+ -                           |ci| among all miniprotocols. To ensure fairness each miniprotocol
+ -                           |cr| can at most have one message in the queue, see
+ -                           +--+ Desired servicing semantics.
+ -                            |
+ -                            V
+ -                         +--+--+ The egress queue is served by a dedicated thread which
+ -                         | mux | chops up the CBOR data into MuxSDUs with at most sduSize
+ -                         +-----+ bytes of data in them.
+ -                            |
+ -                            | MuxSDUs
+ -                            |
+ -                            V
+ -                    +-------+--------+
+ -                    | Bearer.write() | MuxBearer implementation specific write
+ -                    +----------------+
+ -                            |
+ -                            | ByteStrings
+ -                            V
+ -                            o
+ -}
+
+-- | Process the messages from the mini protocols - there is a single
+-- shared FIFO that contains the items of work. This is processed so
+-- that each active demand gets a `maxSDU`s work of data processed
+-- each time it gets to the front of the queue
+mux :: (MonadSTM m, MuxBearer m)
+     => PerMuxSharedState m
+     -> m ()
+mux pmss = do
+    w <- atomically $ readTBQueue $ tsrQueue pmss
+    case w of
+         TLSRDemand mid md d
+             -> processSingleWanton pmss mid md d >> mux pmss
+
+-- | Pull a `maxSDU`s worth of data out out the `Wanton` - if there is
+-- data remaining requeue the `TranslocationServiceRequest` (this
+-- ensures that any other items on the queue will get some service
+-- first.
+processSingleWanton :: (MonadSTM m, MuxBearer m)
+                    => PerMuxSharedState m
+                    -> MiniProtocolId
+                    -> MiniProtocolMode
+                    -> Wanton m
+                    -> m ()
+processSingleWanton pmss mpi md wanton = do
+    maxSDU <- sduSize (bearerHandle pmss)
+    blob <- atomically $ do
+      -- extract next SDU
+      d <- takeTMVar (want wanton)
+      let (frag, rest) = BL.splitAt (fromIntegral maxSDU) d
+      -- if more to process then enqueue remaining work
+      unless (BL.null rest) $
+        do putTMVar (want wanton) rest
```

Comment:
This is important that it's in the same STM transaction as the `takeTMVar` above., otherwise duplex channel could write to it and miss-order bytes.  I would note that, since it makes understanding of the code easier.

Reply:
Yes, accessing the TMVar needs all be from within the same transaction. I'll add something.